Sources/Foundation/Decimal.swift

```diff
@@ -1678,25 +1678,41 @@ public func NSDecimalMultiplyByPowerOf10(_ result: UnsafeMutablePointer<Decimal>
 }
 
 public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> String {
+    let ZERO: UInt8 = 0x30 // ASCII '0' == 0x30
+    let zeroString = String(Unicode.Scalar(ZERO)) // "0"
 
     var copy = dcm.pointee
     if copy.isNaN {
         return "NaN"
     }
     if copy._length == 0 {
-        return "0"
+        return zeroString
     }
 
-    let ZERO: CChar = 0x30 // ASCII '0' == 0x30
-    let DECIMALPOINT: CChar = 0x2e // ASCII '.' == 0x2e
-    let MINUS: CChar = 0x2d // ASCII '-' == 0x2d
-    let bufferSize = 200 // Max value: 39 + 128 + sign + decimal point
-    var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+    var decimalSeparator: String!
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? [NSLocale.Key: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
+    } else if let dictionary = locale as? [String: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
+    }
+    // Defaulting to decimal point
+    if decimalSeparator == nil {
+        decimalSeparator = "."
+    }
+    let decimalSeparatorReversed = String(decimalSeparator.reversed())
```

Comment:
I think the last fix we need is to optimise for the common case that `locale` is `nil` and also not to bother `reverse()`ing the decimal point, this may be a better way to go:

```
let decimalSeparatorReversed: String = {
    guard locale != nil else { return "." }

    var decimalSeparator: String? = nil
    if let locale = locale as? Locale {
        decimalSeparator = locale.decimalSeparator
    } else if let dictionary = locale as? [NSLocale.Key: String] {
        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
    } else if let dictionary = locale as? [String: String] {
        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
    }
    if let ds = decimalSeparator { return String(ds.reversed()) }
    // Defaulting to decimal point
    return "."
}()
```

Also worth adding a test with a mutli-character `locale` to test it is added to the string correctly, using a locale like the following should be sufficient:
```
let locale: [NSLocale.Key: String] = [NSLocale.Key.decimalSeparator: "()"]
```